### ISSUE TITLE
Super weird keyboard input regression

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#2948]
1. Download the repro pack [CPCE.zip](https://github.com/dosbox-staging/dosbox-staging/files/12682685/CPCE.zip) (it contains the free CPCE Amstrad CPC emulator).

2. Unpack & start with `dosbox --noprimaryconf`.

4. Press any key when the about dialog appears at startup.

5. **First issue:** The blue CPC startup screen will appear, but a "down arrow" symbol will be being auto-typed, followed by multiple "right arrow" keys on key repeat. This will only stop if you press some keys (press enter twice).

6. **Second issue:** Press the "A" key – this will type `aX`. Press Space then press the "S" key — this will type `sZ`.
 
The below screenshot illustrates both problems. The emulator works flawlessly with 0.80.1 and SVN. I've only tested on macOS. I'm suspecting yet another SDL regression... I'm using SDL 2.28.2. Would be good if people could report how this behaves on Linux.

<img width="1047" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/21837429-8bdd-4031-83f0-9329290e31c8">

**Comment by johnnovak** [comment]
Ha! Bisected tho this. SDL is off the hook 😅 

```
731547af978e108674f2a482c8ced42fd64dc68a is the first bad commit
commit 731547af978e108674f2a482c8ced42fd64dc68a
Author: Magnus Larsen <golgothasTerror101@gmail.com>
Date:   Mon May 22 06:00:37 2023 -0700

    Create and update joy binds even without physical

    This has been broken since my change (#154) in 2019.

    Emulated joysticks still have issues with key-repeat, but this is a
    seperate issue (was not caused by my 2019 PR)

    See issue #2474

 src/gui/sdl_mapper.cpp |  3 +--
 src/gui/sdlmain.cpp    | 14 ++++++--------
 2 files changed, 7 insertions(+), 10 deletions(-)
```

**Comment by johnnovak** [comment]
The plot thickens: this happens because the buttons of joy 1 are mapping to "A" and "S" in my config, respectively. Still a regression, of course, but that's another piece of the puzzle.

So inserting this snippet into the mapper should be enough to trigger it:

```
jbutton_0_0 "key 4" 
jbutton_0_1 "key 22" 
```

Updating the ticket, plus I think we should have another config option to ignore the mapperfile as random mappings can also throw off testing and troubleshooting.

**Comment by Torinde** [comment]
> another command line option to ignore any mapperfile

FYI:
- just sharing, I know you're not replicating X:
> [-defaultmapper](https://github.com/joncampbell123/dosbox-x/wiki/DOSBox‐X’s-Command‐Line-Options) : Uses the default key mappings for DOSBox-X (do not try to load a mapper file such as mapper-dosbox-x.map).

- #342

**Comment by johnnovak** [comment]
> > another command line option to ignore any mapperfile
> 
> FYI:
> 
>     * just sharing, I know you're not replicating X:
> 
> 
> > [-defaultmapper](https://github.com/joncampbell123/dosbox-x/wiki/DOSBox%E2%80%90X%E2%80%99s-Command%E2%80%90Line-Options) : Uses the default key mappings for DOSBox-X (do not try to load a mapper file such as mapper-dosbox-x.map).
> 
>     * [DOSBox merging joysticks #342](https://github.com/dosbox-staging/dosbox-staging/issues/342)

That's not a bad name. Although for consistency with `noprimaryconf` and `nolocalconf` I think `nomapperconf` would be perhaps better.

**Comment by weirddan455** [comment]
I did some testing on this.  https://github.com/dosbox-staging/dosbox-staging/commit/731547af978e108674f2a482c8ced42fd64dc68a is a bit of a red hearing.  The commit does exactly what it says it does.  It fixes emulated joysticks.  I checkout out the parent commit https://github.com/dosbox-staging/dosbox-staging/commit/2143e19953a8eccd5deb97a92c5c423f526dce7c and it has the same bug when a real joystick is plugged in (tested with an Xbox One controller).

**Comment by johnnovak** [comment]
> I did some testing on this. [731547a](https://github.com/dosbox-staging/dosbox-staging/commit/731547af978e108674f2a482c8ced42fd64dc68a) is a bit of a red hearing. The commit does exactly what it says it does. It fixes emulated joysticks. I checkout out the parent commit [2143e19](https://github.com/dosbox-staging/dosbox-staging/commit/2143e19953a8eccd5deb97a92c5c423f526dce7c) and it has the same bug when a real joystick is plugged in (tested with an Xbox One controller).

Well, but on macOS with that commit I get the weird keyrepeat behaviour, with the one before it's all good. I don't have any joysticks or controllers plugged in.

**Comment by weirddan455** [comment]
>  Well, but on macOS with that commit I get the weird keyrepeat behaviour, with the one before it's all good. I don't have any joysticks or controllers plugged in.

Right and the same thing happens on Linux when I unplug my controller.  https://github.com/dosbox-staging/dosbox-staging/commit/731547af978e108674f2a482c8ced42fd64dc68a simply emulates a joystick being plugged in so now it behaves the same way.

When I have a real joystick plugged in, I get the exact same behavior you mentioned both before and after that commit.  Same arrow patterns and pressing a key on the keyboard or on the gamepad stops the repeat.

The arrows can be controlled by the left stick on my Xbox controller and once you move the stick, the repeat bug starts up again.

It's strange because I see no such behavior in this joytest program.  The X and Y axis stay perfectly centered at 128/128 when I'm not touching the stick (and so does the "emulated" joystick when I unplug my controller).  https://www.vogons.org/viewtopic.php?p=187168#p187168

**Comment by johnnovak** [comment]
> When I have a real joystick plugged in, I get the exact same behavior you mentioned both before and after that commit. Same arrow patterns and pressing a key on the keyboard or on the gamepad stops the repeat.

Ok, so maybe a bug in the CPCE emulator? In any case, it seems like a minor issue (I can just press Enter twice at startup or after a CPC reset with Ctrl+F5).

Btw, the 16-bit real-mode version is also included in the pack (`CPCE.EXE`), and that one doesn't exhibit the problem. Just make sure to set `cycles = 300000` before starting it, otherwise it will take forever at 3000 cycles (it requires a Pentium MMX).

**Comment by kcgen** [comment]
The joystick regression fix, when no joystick is physically plugged,  brought back the presence of a virtual-mappable joystick.. like a dummy ever-present joystick that DOS programs see.  It lets non-joystick owners easily map their keyboard buttons to joystick events and play joystick-exclusive games. 

..  but the presence of a joystick was a bit rare in the DOS days. Perhaps for every DOS that existed, only 1 in 100 had a joystick. I threw away my Gravis due to drift, and all of my fellow DOS PC friends didn't have one. 

Can test the prior behavior with:

```ini
[joystick]
joysticktype = hidden
```

This hides the joystick from DOS's awareness and I/O port eventing.

**Comment by johnnovak** [comment]
> Can test the prior behavior with:
> 
> ```ini
> [joystick]
> joysticktype = hidden
> ```
> 
> This hides the joystick from DOS's awareness and I/O port eventing.

Indeed, that fixes it! Wasn't aware of this setting (I very rarely play games that require a joystick; I'm a keyboard + mouse guy).

**Comment by johnnovak** [comment]
It seems this is not really a bug, more like a weird CPCE issue, plus there's a workaround. CPCE might as well behave the same way on a real computer when a joystick is plugged in for all we know (unfortunately, I can't test it as I don't have an analog joystick for my retro PCs). Happy to close the ticket if you guys agree @kcgen @weirddan455 .

**Comment by weirddan455** [comment]
Yeah I don’t know how to determine if this is our bug or just a quirk of this software short of testing it on real hardware which I don’t have.

Thank you @kcgen for explaining the situation in better words.  That was what I was trying to get at, although I was not aware of that config flag.

**Comment by kcgen** [comment]
I think we just need a confirmed that this isn't a regression versus SVN, which should also exhibit the same issue.

**Comment by johnnovak** [comment]
Ok, so the problem happens with `main` with `cycles = auto` only. If I set `cycles = max` or any fixed cycles value manually, I can't trigger it.

I can't trigger it in SVN or in 0.80.1 either with any cycles settings.

It's a weird issue, but a rather low priority one, so I'll just keep it open until somehow has the appetite to dig deeper because something strange is going on here.
